cip/CIP2016-07-07-Parameter-syntax.adoc

```diff
@@ -0,0 +1,98 @@
+= CIP2016-07-07 - Parameter syntax
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Mats Rydberg <mats@neotechnology.com>
+
+[abstract]
+.Abstract
+--
+This CIP suggests changing syntax for parameter usage, by replacing the curly braces notation (`{}`) with the dollar sign (`$`).
+As curly braces are used by the language for literal maps (in expressions and patterns) and has been suggested for upcoming subquery syntax, their usage as parameter delimiters may lead to confusion when writing or reading queries.
+A single, unique and hitherto unused character used as a prefix increases the readability of parameters when compared with delimiting parameters using overloaded characters. Moreover, this allows for cleaner syntax.
+--
+
+toc::[]
+
+== Motivation
+
+The clause `MATCH (a {prop: {param}})` uses a nested curly brace structure which is not easily readable.
+The alternative `MATCH (a {prop: $param})` does a much better job at signalling the parameter usage to the reader.
+Using a single leading character is a very common way of referencing variables in general programming languages.
+Specifically, the following list of languages use the dollar sign:
+
+- Scala (string interpolation)
+- PHP (variable definition)
+- AutoIt (variable definition)
+- Perl (scalar variables)
+- Ruby (global variables)
+- BASIC (trailing character for variables)
+
+== Proposal
+
+The concrete suggestion made by this CIP is to define the parameter syntax to use a dollar sign.
+This conflicts with the current use of wrapping curly braces, which this CIP suggests removing.
+
+=== Syntax
+
+.Parameter expression:
+[source, ebnf]
+----
+parameter = "$", variable ;
+----
+
+=== Semantics
+
+The semantics are unchanged.
+
+=== Examples
+
+.Using a parameter in a property predicate.
+[source, cypher]
+----
+MATCH (a {prop: $param})
+RETURN count(a)
+----
+
+.Using a parameter in a `WITH` clause.
+[source, cypher]
+----
+WITH $param AS p
+RETURN toString(p)
+----
+
+.Using a parameter with an escaped name in a `WITH` clause.
+[source, cypher]
+----
+WITH $`Der Parameter` AS p
+RETURN toString(p)
+----
+
+.Using a map parameter as a property predicate.
+[source, cypher]
+----
+MATCH (n $param)
```

Comment:
It should be noted that the space is required here, since otherwise it will be taken as a single variable name `n$param`.